### DISCUSSION TITLE
ArrayField operations

### DISCUSSION THREAD

**Discussion by chladog** [payloadcms/payload#615] [Feature Requests & Ideas]
I have long arrays of objects in my ArrayFields and I'm looking for operations similar to MongoDB's $set and find of array item by prop value, as I want to be able to update and get single items without retrieving/sending the full array.
I'm able to use native MongoDB operations, but I would rather use payload's functions to let it handle permissions.

**$push**, **$pull** functions is actually working now:
```JS
              task = {
                  ...
              };
              await payload.update({
                id: args.id,
                collection: "myCollection",
                data: {
                  $push: { tasks: task },
                },
              });
```

For update operation it would ideally take **[arrayFilters](https://www.mongodb.com/docs/manual/reference/method/Bulk.find.arrayFilters/)** property (and rest/graphql param) and would look like this:
```JS
        args.data = { $set: { 'tasks.$[task]': {text: 'task body text', done: false} } };      // coming from graphql args
        args.arrayFilters =  [{ 'task._id': MYTASKID}]; // coming from graphql args, the arrayfilter would be user-defined (client-app)
        payload.update({
          collection: args.collection,
          id: args.id,
          data: args.data,
          arrayFilters: args.arrayFilters
        });
```

I'm not sure how find operation for particular array item should look like, but maybe allow [**$elemMatch**](https://www.mongodb.com/docs/manual/reference/operator/query/elemMatch/) argument for find's ```where``` prop. Then I could create my custom query that would return only matched array item.

**Comment by jmikrut** [comment]
Hey @chladog — because this is so Mongo-specific, I feel like we will have a hard time creating a pattern that will be able to be leveraged and extended on if we decide to support other databases in the future. It almost feels like something that _**should**_ be done with Mongoose / raw Mongo given its specificity. 

I think what would make sense to me is to create your own endpoint or GraphQL mutation. You can leverage Payload's authentication and even its access control in your own logic so ideally you'd be able to replicate whatever you're trying to retain from Payload operations themselves. 

What do you think?

**Reply by chladog** [reply]
Hello @jmikrut ,
I understand your reasoning here. It wouldn't have to be a part of default resolvers. However it would be great to have more control over the parameters that are being passed to db call by **local api**.

Currently in payload.update implementation there is this main mongo-dependent code:
```JS
await Model.findByIdAndUpdate(
        { _id: id },
        result,
        { new: true },       // this is args[2] I'm interested in
      );
``` 
if we get a way to override these default paramters by some kind of **custom filter function** where I could merge my parameters then I would be able to implement mentioned logic in my custom graphql endpoint without the heavy lifting of hooks and access logic (I really don't feel comfortable to reimplement those in every such custom resolver, as it gets big and ugly).
Such function would be **DB-agnostic** as well.

payload.update would just invoke the function right before db call:
```JS
args = dbFilter ? dbFilter(args) : args;
await Model.findByIdAndUpdate(
      ...args
      );
``` 

It would be used like this:
```JS
await payload.update({
  id: args.weddingId,
  collection: "myCollection",
  data: {
    $set: { "tasks.$": incomingTask },
  },
  dbFilter: (args) => {
    Object.assign(args[2], {
      arrayFilter: [{ "task._id": incomingTaskId }],
    });
    return args;
  },
});
```

Same pattern could be applied to all other operations, create, find, delete...
